### PR TITLE
Implementations for remaining async cache methods (initialize, invalidate, remove)

### PR DESCRIPTION
**Pull request by sphill99**
Implemented the remaining methods for this class. This involved adding read methods that took in an AsyncFileChannel. I used a Future for these operations in place of a CompletionHandler, as that seemed much simpler in this case. Consolidated the callback interface for put, init, invalidate, and remove, since they all did the same thing. Can change this back to them each having their own if there are any benefits to that.

Initialize still has a loop over a files array, but couldn't figure out how to do that asynchronously since our map is not thread safe.

### CODE REVIEW

File: src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java [hunk 1]
```diff
@@ -37,6 +48,104 @@ static String getFilenameForKey(String key) {
...
+            totalSize += entry.size;
+        } else {
+            CacheHeader oldEntry = entries.get(key);
+            totalSize += (entry.size - oldEntry.size);
+        }
+        entries.put(key, entry);
+        return totalSize;
+    }
+
+    /** Removes the entry identified by 'key' from the cache. This modifies the entries map */
```
Comment (sophiahaim) [inline src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:140]: Add @return The updated totalSize

File: src/main/java/com/android/volley/toolbox/DiskBasedCache.java [hunk 2]
```diff
@@ -255,31 +256,18 @@ public synchronized void put(String key, Entry entry) {
     /** Removes the specified key from the cache if it exists. */
     @Override
     public synchronized void remove(String key) {
-        boolean deleted = getFileForKey(key).delete();
+        boolean deleted = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier).delete();
         removeEntry(key);
```
Comment (sophiahaim) [inline src/main/java/com/android/volley/toolbox/DiskBasedCache.java:260]: Can you use the new DiskBasedCacheUtility#removeEntry function here?

File: src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java [hunk 3]
```diff
@@ -136,4 +329,39 @@ static void writeHeaderList(List<Header> headers, OutputStream os) throws IOExce
...
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    static List<Header> readHeaderList(AsynchronousFileChannel channel, int offset)
+            throws IOException {
+        int size = readInt(channel, offset);
+        if (size < 0) {
+            throw new IOException("readHeaderList size=" + size);
+        }
+        offset += 4;
+        List<Header> result =
+                (size == 0) ? Collections.<Header>emptyList() : new ArrayList<Header>();
```
Comment (sophiahaim) [inline src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:342]: nit: don't need the parentheses

File: src/main/java/com/android/volley/toolbox/DiskBasedCache.java (hunk 2 above)
Comment (sphill99) [reply src/main/java/com/android/volley/toolbox/DiskBasedCache.java:260]: Oh yeah. It looks like I didn't make those changes for DiskBasedCache on this branch.

Comment (joebowbeer) [reply src/main/java/com/android/volley/toolbox/DiskBasedCache.java:260]: We should hold off reviewing this PR until the other PR on which it is based is merged?

Comment (sphill99) [reply src/main/java/com/android/volley/toolbox/DiskBasedCache.java:260]: Yeah, I think that's smart

**Comment by sphill99** [comment]
I didn't realize how much overlap there was between this PR and the PutImpl one. So I'm making this a draft for now.

**Pull request closed without merging**